Add col2im_batched kernel

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84543

Closes #84407

This changes col2im on CUDA to launch a single batch-aware kernel
instead of launching n single slice kernels.

The `istft` call in the linked issue goes from 98.7 ms to 858 us on my
machine, for an over 100x speedup.

peterbell10 committed:
Add col2im_batched kernel

Closes #84407

This changes col2im on CUDA to launch a single batch-aware kernel
instead of launching n single slice kernels.

The `istft` call in the linked issue goes from 98.7 ms to 858 us on my
machine, for an over 100x speedup.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add col2im_batched kernel"


Closes #84407

This changes col2im on CUDA to launch a single batch-aware kernel
instead of launching n single slice kernels.

The `istft` call in the linked issue goes from 98.7 ms to 858 us on my
machine, for an over 100x speedup.

[ghstack-poisoned]

peterbell10 committed:
Update on "Add col2im_batched kernel"


Closes #84407

This changes col2im on CUDA to launch a single batch-aware kernel
instead of launching n single slice kernels.

The `istft` call in the linked issue goes from 98.7 ms to 858 us on my
machine, for an over 100x speedup.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/Col2Im.cu:
```diff
@@ -119,28 +116,26 @@ void col2im_out_cuda_template(
-      input_n = input.select(0, elt);
-      output_n = output.select(0, elt);
-
-      col2im<scalar_t, accscalar_t>(
```
is col2im/col2im_kernel still used anywhere, or can they be removed?

peterbell10 replied:
Yes, It's used here for example:
https://github.com/pytorch/pytorch/blob/fa99b7b8f726f8ce63f0dff076b7e9171e3dd40a/aten/src/ATen/native/cuda/ConvolutionMM2d.cu#L281

peterbell10 committed:
Retrigger CI on "Add col2im_batched kernel"


Closes #84407

This changes col2im on CUDA to launch a single batch-aware kernel
instead of launching n single slice kernels.

The `istft` call in the linked issue goes from 98.7 ms to 858 us on my
machine, for an over 100x speedup.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

ixNeo commented:
hi, which torch version support this speed behavior
